src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -70,7 +70,72 @@ public sealed partial class JsonSerializerOptions
             return converters;
 
             void Add(JsonConverter converter) =>
-                converters.Add(converter.TypeToConvert!, converter);
+                converters.Add(converter.TypeToConvert, converter);
+        }
+
+        internal JsonConverter GetDictionaryKeyConverter(Type keyType)
+        {
+            _dictionaryKeyConverters ??= GetDictionaryKeyConverters();
```

Comment:
fwiw - can we fetch from `s_defaultSimpleConverters` instead and not have this cache? There might be some complexity where we exclude the 5 unsupported types from the lookup.

Reply:
If we remove the cache, which also works as an allow-list, we would have to add some other mechanism to detect if the type is supported as dictionary key.

That can be: 
* Have a set of supported types (similar to how the current cache is used now).
* Add a bool property to JsonConverter<T> to indicate if the type `CanBeDictionaryKey`.

I think the former is cleaner and since we still need to cache for the `Enum` case, it doesn't make too much sense to me to remove it.